ambry-server/src/main/java/com/github/ambry/server/StatsManager.java

```diff
@@ -435,53 +478,108 @@ long getPermanentDeleteTombstoneSize() {
   }
 
   /**
-   * Runnable class that collects, aggregate and publish stats via methods in StatsManager.
+   * Runnable class that collects, publish account stats to mysql database and local backup file.
    */
-  private class StatsAggregator implements Runnable {
+  class AccountStatsPublisher implements Runnable {
     private volatile boolean cancelled = false;
     private final AccountStatsMySqlStore accountStatsMySqlStore;
 
-    StatsAggregator(AccountStatsMySqlStore accountStatsMySqlStore) {
+    AccountStatsPublisher(AccountStatsMySqlStore accountStatsMySqlStore) {
       this.accountStatsMySqlStore = accountStatsMySqlStore;
     }
 
     @Override
     public void run() {
-      logger.info("Aggregating stats for local report");
-      try {
-        resetDeleteTombstoneStats();
-        long totalFetchAndAggregateStartTimeMs = time.milliseconds();
-        StatsSnapshot aggregatedSnapshot = new StatsSnapshot(0L, new HashMap<>());
-        List<PartitionId> unreachablePartitions = new ArrayList<>();
-        Iterator<PartitionId> iterator = (new HashSet<>(partitionToReplicaMap.keySet())).iterator();
-        while (!cancelled && iterator.hasNext()) {
-          PartitionId partitionId = iterator.next();
-          logger.info("Aggregating stats for local report started for store {}", partitionId);
-          collectAndAggregate(aggregatedSnapshot, partitionId, unreachablePartitions);
-        }
-        aggregatedSnapshot.updateValue();
-        List<String> unreachableStores = examineUnreachablePartitions(unreachablePartitions);
-        updateAggregatedDeleteTombstoneStats();
-        metrics.initDeleteStatsGaugesIfNeeded();
-        if (!cancelled) {
-          metrics.totalFetchAndAggregateTimeMs.update(time.milliseconds() - totalFetchAndAggregateStartTimeMs);
-          StatsHeader statsHeader = new StatsHeader(StatsHeader.StatsDescription.STORED_DATA_SIZE, time.milliseconds(),
-              partitionToReplicaMap.keySet().size(), partitionToReplicaMap.keySet().size() - unreachableStores.size(),
-              unreachableStores);
-          // First write to account stats
-          StatsWrapper statsWrapper = new StatsWrapper(statsHeader, aggregatedSnapshot);
+      Object lock = accountStatsMySqlStore != null ? accountStatsMySqlStore : this;
```

Comment:
Probably simpler to use synchronized(this) in both cases.

Reply:
I want to synchronize on accountStatsMysqlStore since accountStatsMysqlStore only has one connection and multiple threads accessing this connection is not safe. That's why every time I want to use accountStatsMysqlStore, I would have to lock this object.
There are other threads trying to use accountStatsMysqlStore, one of them is MysqlReportAggregatorTask. This is a aggregation task that read all the stats reports and aggregate them and then save the aggregated result back to mysql database. By synchronizing on accountStatsMysqlStore,  we can be sure that there is no concurrency issue with accessing the single connection in accountStatsMysqlStore.
However, the accountStatsMysqlStore might be null, it's not yet enabled everywhere.

Reply:
Okay, makes sense.